Add Debian Stretch support

smortex opened the pull request:
Debian Stretch (9) ships with OpenSSL 1.1.0, while previous versions
where using OpenSSL 1.0.0.  Elastic maintains two repository for
handling this situation.

More info:
https://www.elastic.co/guide/en/elasticsearch/client/curator/5.3/apt-repository.html

Debian versions prior to 8 are obsolete and not supported.
Future Debian versions are currently not supported, and adding support
should be straightforward when they are released.

smortex committed:
Add Debian Stretch support

Debian Stretch (9) ships with OpenSSL 1.1.0, while previous versions
where using OpenSSL 1.0.0.  Elastic maintains two repository for
handling this situation.

More info:
https://www.elastic.co/guide/en/elasticsearch/client/curator/5.3/apt-repository.html

Debian versions prior to 8 are obsolete and not supported.
Future Debian versions are currently not supported, and adding support
should be straightforward when they are released.

smortex committed:
Update metadata.json's supported Debian versions

smortex committed:
Add modules extracted from Puppet core in 6.0

smortex committed:
Ubuntu is a kind of Debian

No version of Ubuntu seems to currently ships OpenSSL 1.1.0.

smortex committed:
Honor PUPPET_VERSION

It's extensively used in Travis-CI configuration but has no effect.

smortex committed:
Update the build matrix

Puppet 4.x has been EOL for a few month.  Test against puppet 5.x and
6.x.

smortex committed:
Sync supported versions of Ubuntu

* Add 18.04 which is part of the test suite and was missing;
* Remove 10.04 which has reached EOL.

smortex committed:
Fix deployment job by adding pdk to the Gemfile

smortex commented:
After a few trial & errors, this seems to be building fine on Travis-CI \o/

Ready for review!

mvisonneau commented on manifests/repository/apt.pp:
```diff
@@ -5,15 +5,28 @@
     class { '::apt': }
   }
 
+  $repo_suffix = $debian_major_version ? {
```
I believe you can rely straight on the fact, I don't see any use case that would require updating it.

smortex replied:
The "new" repo is "debian9" while the previous one is "debian".  As far as I can see, "debian7", "debian8", etc are not valid, hence the extra variable.

mvisonneau commented on manifests/repository/apt.pp:
```diff
@@ -5,15 +5,28 @@
+  $repo_suffix = $debian_major_version ? {
+    # Debian releases
+    '8'     => '',
+    '9'     => '9',
```
This is probably easier :
```
$repo_suffix = $::os['release']['major'] ? {
  '9'         => '9',
  default => '',
}
```

smortex replied:
I chose to do that way because Debian 10 will surely not work with the "debian" repo.  It _might_ work with the "debian9" repo, but maybe not :man_shrugging: 

It might be better to crash hard with unsupported systems, and add support for them accordingly when they are released?

mvisonneau replied:
Ah ok, yes in that case let's leave the parameter `$debian_major_version` so that people can potentially use it with debian 10 before we patch the module 👍. However lets keep only *9* and *default* in the case statement

smortex replied:
Added in b41fee3cf51abc25453f5717728f40aa9b10b396

Please note that I am not endorsing this since the module will successfully deploy a broken setup, and because of the way the module works, the end-user will not be aware of the breakage (until he discovers that curator is non-functional because he ran out of disk space).

For the record, here is what the problem looks like when you hit it: https://github.com/elastic/curator/issues/1241

mvisonneau replied:
if you prefer I'm happy to get what you used to have but make it fail on default then. Otherwise we're just going to get an undef value. Might also need to get a conditional around it to only execute the switch case on Debian base hosts though.

smortex replied:
Hehe, when no default is present and no match is found, the selector fails (this is what I was relying on):

```
romain@zappy ~ % puppet apply -e '$x = $nobody_knows ? { 1 => "a", 2 => "b", } '                                                                            
Warning: Unknown variable: 'nobody_knows'. (line: 1, column: 6)
Error: Evaluation Error: No matching entry for selector parameter with value '' (line: 1, column: 6) on node zappy.home.sigabrt.org
```

You are right, a default with a better error message is probably the best!

mvisonneau commented on metadata.json:
```diff
@@ -38,18 +38,17 @@
     {
```
Could you please also add the new Puppet versions being supported? 🙏

smortex replied:
Ooops, missed that!

1200aeeb21a339a22b2b3b5f8ef98f12c592afcb

smortex replied:
Build is failing.

I'm afraid that supporting the module on an unsupported version of Puppet running on an unsupported version of Ruby is a waste of resources :confused:

mvisonneau commented on .travis.yml:
```diff
@@ -12,16 +12,18 @@ script:
-    env: PUPPET_VERSION="~> 4.0" CHECK=metadata_lint
-  - rvm: 2.4.0
-    env: PUPPET_VERSION="~> 4.0" CHECK=build DEPLOY_TO_FORGE=yes
+  - rvm: 2.5.1
```
could we keep support/tests for Puppet 4.x please

smortex replied:
8556db2124339f131f8becb3596b417a64f4e9f6

smortex replied:
Build is failing.

I'm afraid that supporting the module on an unsupported version of Puppet running on an unsupported version of Ruby is a waste of resources :confused:

smortex replied:
To complete information:

* [The latest version of Puppet 4.x (4.10.2)  ships in PE 2016.4.15 with Ruby 2.1.9](https://puppet.com/docs/pe/2019.0/component_versions_in_recent_pe_releases.html)
* [PE 2016.4.15 has reached EOL in October 2018](https://groups.google.com/forum/#!topic/puppet-users/78ZcVVyuIMQ)
* [Ruby 2.1.x has reached EOL in March 2017](https://en.wikipedia.org/wiki/Ruby_(programming_language)#Table_of_versions)

mvisonneau replied:
ah right, fair enough thanks for the info!

smortex replied:
Fine, removing 8556db2124339f131f8becb3596b417a64f4e9f6

mvisonneau commented on .fixtures.yml:
```diff
@@ -1,6 +1,12 @@
 fixtures:
   repositories:
     apt: "git://github.com/puppetlabs/puppetlabs-apt.git"
+    cron_core:
```
are those modules new requirements ? if so, please add them to the `metadata.json`

smortex replied:
Not really, Puppet 6.x has removed some built-in modules which are now provided as third-party modules.

However, the puppet-agent package ships with these modules, so you just have to deal with them in the test suite like here.

mvisonneau commented on .fixtures.yml:
```diff
@@ -1,6 +1,12 @@
+      repo: "https://github.com/puppetlabs/puppetlabs-cron_core"
+      puppet_version: '>= 6.0.0'
     stdlib: "git://github.com/puppetlabs/puppetlabs-stdlib.git"
+    yumrepo_core:
```
do we need this module if we only support Debian based instances ? 🤔

smortex replied:
Since the module use `yumrepo` in `manifests/repository/yum.pp`, and `yumrepo` is no longer a built-in module, I had to add this too.  Same remark as above: puppet-agent ships with the module, so there is no need to update `metadata.json`.

mvisonneau reviewed (changes requested):
Thanks a lot for your contribution guys, awesome work. Few things I would like to get polished but otherwise LGTM 👍

smortex committed:
We support Puppet up to 6.x

smortex committed:
Do not white-list supported Debian-based distros

The maintainer prefers that the module does not fail hard when running
on an unsupported Linux distro.  Default to use the OpenSSL 1.0.0
repository expect from Debian 9 where OpenSSL 1.1.0 is installed.

mvisonneau commented:
Thanks a lot @smortex ! 🙏
